src/cache.ts

```diff
@@ -83,31 +83,34 @@ function findPackageManager(id: string): PackageManager {
 /**
  * A function that generates a cache key to use.
  * Format of the generated key will be "${{ platform }}-${{ id }}-${{ fileHash }}"".
- * If there is no file matched to {@link PackageManager.path}, the generated key ends with a dash (-).
  * @see {@link https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#matching-a-cache-key|spec of cache key}
  */
-async function computeCacheKey(packageManager: PackageManager) {
-  const hash = await glob.hashFiles(packageManager.pattern.join('\n'));
-  return `${CACHE_KEY_PREFIX}-${process.env['RUNNER_OS']}-${packageManager.id}-${hash}`;
+async function computeCacheKey(
+  packageManager: PackageManager,
+  cacheDependencyPath: string
+) {
+  const pattern = cacheDependencyPath
+    ? cacheDependencyPath.trim().split('\n')
+    : packageManager.pattern;
+  const fileHash = await glob.hashFiles(pattern.join('\n'));
+  if (!fileHash) {
+    throw new Error(
+      `No file in ${process.cwd()} matched to [${pattern}], make sure you have checked out the target repository`
+    );
+  }
+  return `${CACHE_KEY_PREFIX}-${process.env['RUNNER_OS']}-${packageManager.id}-${fileHash}`;
 }
 
 /**
  * Restore the dependency cache
  * @param id ID of the package manager, should be "maven" or "gradle"
+ * @param cacheDependencyPath The path to a dependency file
  */
-export async function restore(id: string) {
+export async function restore(id: string, cacheDependencyPath: string) {
   const packageManager = findPackageManager(id);
-  const primaryKey = await computeCacheKey(packageManager);
-
+  const primaryKey = await computeCacheKey(packageManager, cacheDependencyPath);
   core.debug(`primary key is ${primaryKey}`);
   core.saveState(STATE_CACHE_PRIMARY_KEY, primaryKey);
-  if (primaryKey.endsWith('-')) {
-    throw new Error(
-      `No file in ${process.cwd()} matched to [${
-        packageManager.pattern
-      }], make sure you have checked out the target repository`
-    );
-  }
```

Comment:
We should keep this, as the post-step of the action, namely save() function depends on the state `STATE_CACHE_PRIMARY_KEY`. If you throw error inside the computeCacheKey() function, this state will not be set.

Reply:
I'd like to know why that's necessary before reverting. When there is no gradle (or maven, sbt) file to compute the cache key, we should not save anything on post process, not just restore something. So in that situation, and if `save()` is called, we should make it quickly return, now it does with `Error retrieving key from state.`. But before that, `save()` is never called in this situation; only called on successful job checked in `saveCache()`. Why do you think invalid primary key state should be saved?

Reply:
I agree with you, post-step has `isJobStatusSuccess()` called inside that means that saving state is not required.